xhr/send-authentication-competing-names-passwords.htm

```diff
@@ -5,50 +5,41 @@
     <script src="/resources/testharness.js"></script>
     <script src="/resources/testharnessreport.js"></script>
     <script src="/common/utils.js"></script>
-    <link rel="help" href="https://xhr.spec.whatwg.org/#the-open()-method" data-tested-assertations="following::ol[1]/li[9]/ol[1]/li[1] following::ol[1]/li[9]/ol[1]/li[2]" />
-    <link rel="help" href="https://xhr.spec.whatwg.org/#the-send()-method" data-tested-assertations="following::code[contains(@title,'http-authorization')]/.." />  </head>
+  </head>
   <body>
     <div id="log"></div>
     <script>
       function request(user1, pass1, user2, pass2, name) {
-        // user1, pass1 will if given become userinfo part of URL
-        // user2, pass2 will if given be passed to open() call
         test(function() {
-          var client = new XMLHttpRequest(),
-              urlstart = "", userwin, passwin
-          // if user2 is set, winning user name and password is 2
-          if(user2)
-            userwin = user2, passwin = pass2
-          // if user1 is set, and user2 is not set, user1 and pass1 win
-          if(user1 && ! user2)
-            userwin = user1, passwin = pass1
-          // if neither user name is set, pass 2 wins (there will be no userinfo in URL)
-          if (!(user1 || user2))
-            passwin = pass2
+          const client = new XMLHttpRequest(),
+                userwin = user2 || user1,
+                passwin = pass2 || pass1;
+          let urlstart = "";
           if(user1) { // should add userinfo to URL (there is no way to create userinfo part of URL with only password in)
-            urlstart = "http://" + user1
-            if(pass1)
-              urlstart += ":" + pass1
-            urlstart += "@" + location.host + location.pathname.replace(/\/[^\/]*$/, '/')
+            urlstart = "http://" + user1;
+            if(pass1) {
+              urlstart += ":" + pass1;
+            }
+            urlstart += "@" + location.host + location.pathname.replace(/\/[^\/]*$/, '/');
           }
```

Comment:
else {
  assert_equals(null, pass2, 'when user1 is null then pass1 must be null');
}
?

Reply:
I don't quite understand this suggestion. This is just about setting up the URL.

Reply:
Sorry, that should be assert_equals(null, pass1, 'when user1 is null then pass1 must be null');

The test will be broken when user1 is null (or empty) and pass1 is not, as pass1 is set to the url only when user1 is not null (or empty).

Perhaps we could construct a URL in a less hacky way, such as 

const url = new URL('resources/authentication.py', location.href);
if (user1) {
  url.username = user1;
}
if (pass1) {
  url.password = pass1;
}

Reply:
Then, how about this?

```
if (user1 || pass1) {
  urlstart = "http://";
  if (user1)
    urlstart += user1;
  if (pass1)
    urlstart += ":" + pass1;
  urlstart += "@" + location.host + location.pathname.replace(/\/[^\/]*$/, '/');
}
```

Reply:
Yeah, I'll make that change. It does seem incorrect that it states you need user if you've been given pass.

Reply:
I should probably add an additional test for that...